sbomtool/internal/commands/filter/filtering.go

```diff
@@ -0,0 +1,351 @@
+	"github.com/bottlerocket-os/bottlerocket-sdk/sbomtool/go/internal/processor"
+)
+
+// FilteringResult contains the complete results of the SBOM filtering operation.
+// It includes the filtered SBOM, categorized packages, and statistics
+// about the filtering process.
+type FilteringResult struct {
+	FilteredSBOM     *sbom.SBOM
+	RootPackages     []pkg.Package
+	IncludedPackages []pkg.Package
+	ExcludedPackages []pkg.Package
+	Statistics       FilteringStatistics
+}
+
+// FilteringStatistics provides metrics about the filtering operation.
+// It includes counts of input/output packages and relationships, performance timing,
+// and details about root matches and transitive inclusions.
+type FilteringStatistics struct {
+	InputPackages        int
+	OutputPackages       int
+	InputRelationships   int
+	OutputRelationships  int
+	FilteringDuration    time.Duration
+	RootMatches          int
+	TransitiveInclusions int
+}
+
+// FilterSBOMByBuildroot filters an SBOM based on buildroot contents with transitive dependency resolution.
+func FilterSBOMByBuildroot(
+	inputSBOM *sbom.SBOM,
+	buildrootPath string) (*FilteringResult, error) {
+
+	startTime := time.Now()
+
+	slog.Info("Starting SBOM filtering with Syft relationships",
+		"input_packages", inputSBOM.Artifacts.Packages.PackageCount(),
+		"input_relationships", len(inputSBOM.Relationships),
+		"buildroot_path", buildrootPath)
+
+	// Bottlerocket is about 15 layers deep, so 30 will get everything but prevents infinite loops
+	buildrootScanner := buildroot.NewScanner(30, []string{"*.tmp", "*.log"})
+	buildrootResult, err := buildrootScanner.ScanDirectory(buildrootPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to scan buildroot: %w", err)
+	}
+
+	normalizedPaths := buildrootResult.GetNormalizedPaths()
+
+	slog.Debug("Buildroot scanning completed",
+		"total_files", len(buildrootResult.AllFiles),
+		"normalized_paths", len(normalizedPaths))
+
+	rootPackages, err := findPackagesWithBuildrootFiles(
+		inputSBOM.Artifacts.Packages, normalizedPaths)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find root packages: %w", err)
+	}
+
+	slog.Info("Root packages identified",
+		"root_count", len(rootPackages),
+		"buildroot_files", len(normalizedPaths))
+
+	includedPackages, err := resolveTransitiveDependencies(inputSBOM, rootPackages)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve transitive dependencies: %w", err)
+	}
+
+	slog.Info("Transitive dependency resolution completed",
+		"root_packages", len(rootPackages),
+		"total_included", len(includedPackages))
+
+	filteredSBOM, err := createFilteredSBOM(inputSBOM, includedPackages)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create filtered SBOM: %w", err)
+	}
+
+	if err := validateFilteredSBOM(filteredSBOM); err != nil {
+		return nil, fmt.Errorf("filtered SBOM validation failed: %w", err)
+	}
+
+	allPackages := inputSBOM.Artifacts.Packages.Sorted()
```

Comment:
is this sort necessary? in `getExcludedPackages`, we compare against a set, so the order of items in this array shouldn't impact that comparison to construct the `excludedPackages`